rewrite-core/src/main/java/org/openrewrite/GitRemote.java

```diff
@@ -47,69 +48,116 @@ public enum Service {
     }
 
     public static class Parser {
-        private final Map<String, Service> origins;
+        private final List<RemoteServer> servers;
 
         public Parser() {
-            origins = new LinkedHashMap<>();
-            origins.put("github.com", Service.GitHub);
-            origins.put("gitlab.com", Service.GitLab);
-            origins.put("bitbucket.org", Service.BitbucketCloud);
-            origins.put("dev.azure.com", Service.AzureDevOps);
-            origins.put("ssh.dev.azure.com", Service.AzureDevOps);
+            servers = new ArrayList<>();
+            servers.add(new RemoteServer(Service.GitHub, "github.com", URI.create("https://github.com"), URI.create("ssh://git@github.com")));
+            servers.add(new RemoteServer(Service.GitLab, "gitlab.com", URI.create("https://gitlab.com"), URI.create("ssh://git@gitlab.com")));
+            servers.add(new RemoteServer(Service.BitbucketCloud, "bitbucket.org", URI.create("https://bitbucket.org"), URI.create("ssh://git@bitbucket.org")));
+            servers.add(new RemoteServer(Service.AzureDevOps, "dev.azure.com", URI.create("https://dev.azure.com"), URI.create("ssh://git@ssh.dev.azure.com")));
         }
 
-        public Parser registerRemote(Service service, String origin) {
-            if (origin.startsWith("https://") || origin.startsWith("http://") || origin.startsWith("ssh://")) {
-                origin = new Parser.HostAndPath(origin).concat();
+        public URI toUri(GitRemote remote, boolean ssh) {
+            URI selectedBaseUrl;
+
+            if (remote.service == Service.Unknown) {
+                if (PORT_PATTERN.matcher(remote.origin).find()) {
+                    throw new IllegalArgumentException("Unable to guess protocol port combination for an unregistered origin with a port: " + remote.origin);
```

Comment:
nit: I'd prefer less casual phrasing than "guess". Perhaps extract or ascertain.... 

`"Unable to ascertain protocol/port combination for unregistered origin " + remote.origin`

Reply:
Updated to use "determine". which I like a bit better.